feat(backup): add Optimize backup docs (#1348)
camunda/camunda-platform-docs: docs/self-managed/backup-restore/optimize-backup.md

---
id: optimize-backup
title: Backup and restore Optimize data
description: "How to perform a backup of Optimize data and restore the backup."
---

> Optimize stores its data over multiple indices in Elasticsearch. To ensure data integrity across indices, a backup of Optimize data consists of several Elasticsearch snapshots containing sets of Optimize indices. Each backup is identified by a backup ID. A backup with id `backup1` consists of the following Elasticsearch snapshots:

Reviewer: ```suggestion
Optimize stores its data over multiple indices in Elasticsearch. To ensure data integrity across indices, a backup of Optimize data consists of two Elasticsearch snapshots, each containing different sets of Optimize indices. Each backup is identified by a backup ID. For example, a backup with id `backup1` consists of the following Elasticsearch snapshots:
```
Maybe not better, just thinking about being explicit

```
camunda_optimize_backup1_3.9.0_part_1_of_2
camunda_optimize_backup1_3.9.0_part_2_of_2
```

> Optimize provides a simple API to trigger a backup and retrieve information about a given backup's state. During backup creation Optimize can continue running. The backed up data can later be restored using the standard Elasticsearch snapshot restore API.

Reviewer: do you think linking to the restore API documentation makes sense here?

## Prerequisites

The following prerequisites must be set up before using the backup API:

1. A snapshot repository of your choice must be registered with Elasticsearch
2. The repository name must be specified using the `CAMUNDA_OPTIMIZE_BACKUP_REPOSITORY_NAME` environment variable or by adding it to your Optimize configuration:

```yaml
backup:
  repositoryName: <repository name>
```

## Create backup API

> Please note that the backup API can be reached via the `/actuator` management port, by default this is `8092`.  

Reviewer: ```suggestion
Please note that the backup API can be reached via the `/actuator` management port, which by default is `8092`.  
```

Reviewer: Also, can this be changed? It feels like saying "by default" implies the possibility of change, but I don't see this in the docs

Author: We have a ticket to make this configurable.

> The following endpoint can be used to asynchronously trigger the backup process:

Reviewer: ```suggestion
The following endpoint can be used to trigger the backup process:
```
I'm not sure we need this implementation detail. For a user it's quite clear that the trigger happens, and the state can be checked, implying asynchronicity (is that a word?). Wdyt?

```
POST actuator/backup
{
  "backupId": <backupId>
}
```

### Response

| Code             | Description                                                                                                                                                                                                                                 |
| ---------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | 200 OK           | Backup was successfully started, snapshots will be created asynchronously. List of snapshot names is returned in response body (see example below). This list must be persisted together with the backup ID to be able to restore it later. |

Reviewer: the actual creation is synchronous isn't it? i.e. the second backup is _always_ created after the first?

> | 200 OK           | Backup was successfully started, snapshots will be created asynchronously. List of snapshot names is returned in response body (see example below). This list must be persisted together with the backup ID to be able to restore it later. |

Reviewer: > "This list must be persisted together with the backup ID to be able to restore it later"

I'm not sure I understand this sentence, can you elaborate?

Author: Customers must keep track of their backup ID if they want to later use the backup ID for eg finding which snapshots to restore

Author: Will rephrase -- can you also add this comment on the other PRs?

> | 400 Bad Request  | Indicates issues with the request, for example when a backup with the same `backupId` already exists.                                                                                                                                       |

Reviewer: Is it worth saying that the description returned should give more detail?

> | 500 Server Error | All other errors, e.g. issues communicating with Elasticsearch for snapshot creation.                                                                                                                                                       |

Reviewer: is this missing a 409?

### Example request

```

> curl --request POST 'http://localhost:8092/actuator/backup' \

Reviewer: Do we typically include `localhost` in our examples, or rather use placeholders like `<server>`? I don't know the answer to this

Author: I've kept to the same standard as the other backup docs which do have this in the example request, given this is a concrete example and not the generalised endpoint into I would say it makes sense, but I dont mind adjusting either.

Reviewer 2: This isn't standardized yet across our docs.

-H 'Content-Type: application/json' \
-d '{ "backupId": "backup1" }'
```

### Example response

```json
{
  "scheduledSnapshots": [
    "camunda_optimize_backup1_3.9.0_part_1_of_2",
    "camunda_optimize_backup1_3.9.0_part_2_of_2"
  ]
}
```

## Check backup state API

Please note that the backup API can be reached via the `/actuator` management port, by default this is `8092`.  

> As the backup is created asynchronously, the current state of the backup can be checked by calling the following endpoint:

Reviewer: Same comment as above re default

```
GET actuator/backup/{backupId}
```

### Response

| Code             | Description                                                                                 |
| ---------------- | ------------------------------------------------------------------------------------------- |
| 200 OK           | Backup state could be determined and is returned in the response body (see example below)   |
| 404 Not Found    | Backup with given id does not exist.                                                        |
| 500 Server Error | All other errors, e.g. issues communicating with Elasticsearch for snapshot state retrieval |

### Example request

```

> curl 'http://localhost:8092/actuator/backup/backup1'

Reviewer: we should include the method in this request

Author: Can you also add this comment to the other PRs please so they can adjust the same

```

### Example response

```json
{
  "State": "COMPLETED"
}
```

Possible states of the backup:

- `COMPLETED`: The backup can be used for restoring data.
- `IN_PROGRESS`: The backup process for this backup ID is still in progress.
- `FAILED`: Something went wrong when creating this backup. To find our the exact problem, one can use [Elasticsearch Get snapshot status API](https://www.elastic.co/guide/en/elasticsearch/reference/current/get-snapshot-status-api.html) for each of the snapshots included in given backup.
- `INCOMPATIBLE`: The backup is incompatible with current Elasticsearch version.
- `INCOMPLETE`: The backup is incomplete (this could occur when backup process was interrupted or individual snapshots were deleted).

## Restore backup

> There is no Optimize API to perform the backup restore, instead the standard [Elasticsearch restore snapshot API](https://www.elastic.co/guide/en/elasticsearch/reference/current/restore-snapshot-api.html) can be used.

Reviewer: 1. Now I've seen this you can probably ignore my previous comment about linking
2. This link points to the latest ES docs (which is version 8.x), something we don't currently support. Do we have a convention for docs linking?

Reviewer 2: Looks like this would be the correct link for us - https://www.elastic.co/guide/en/elasticsearch/reference/7.17/restore-snapshot-api.html

I can update this in a few minutes unless someone beats me to it 😄

Reviewer 2: This is done.

:::note
Optimize must NOT be running while a backup is being restored.
:::

To restore an existing backup, all the snapshots this backup contains (as listed in the response of the [create backup API request](#example-response)) must be restored using the Elasticsearch API.

To restore a given backup, the following steps must be performed:

1. Stop Optimize.
2. Make sure there are no Optimize indices are present in Elasticsearch (otherwise the restore process will fail).
3. Iterate over all Elasticsearch snapshots included in the desired backup and restore them using the Elasticsearch restore snapshot API.
4. Start Optimize.

Example Elasticsearch request:

```
curl --request POST `http://localhost:9200/_snapshot/repository_name/camunda_optimize_backup1_3.9.0_part_1_of_2/_restore?wait_for_completion=true`

> ```

Reviewer: General comment about this file - Can you elaborate on what the Optimize version compatibility would be? i.e. can a customer restore a backup from 3.9.0 on Optimize 3.10.0? Or compatibility between patch versions for example?

> ```

Reviewer: General comment about this file - Do we need to say that this feature is experimental? Or will that be included in the general doc rather than the Optimize one?

Author: I actually added a note at the top which I didn't push 🙃 , will include in the next commit
